### PR TITLE
feat(admin): support partition validation and replica assignments

### LINKED ISSUE #3035: feat(admin): expose partition expansion validation and replica assignments
## Problem

[CreatePartitionsAsync](https://github.com/thomhurst/Dekaf/blob/7d10d1cbf23c081204d4ce351b597f76d76f5156/src/Dekaf/Admin/AdminClient.cs#L2294) accepts only topic-to-count mappings. The wire [CreatePartitionsRequest](https://github.com/thomhurst/Dekaf/blob/7d10d1cbf23c081204d4ce351b597f76d76f5156/src/Dekaf/Protocol/Messages/CreatePartitionsRequest.cs#L7) already supports ValidateOnly and explicit replica assignments, but neither can be selected through the public admin API.

Applications cannot validate a partition expansion without applying it, or place new partition replicas directly on selected brokers through this operation.

## Scope and acceptance criteria

- Add an additive typed partition-expansion specification with total partition count and optional replica assignments for newly added partitions.
- Add options exposing ValidateOnly and operation timeout; retain the existing count-only overload.
- Validate assignment shape and preserve preferred-replica ordering. Do not alter assignments of existing partitions.
- Ensure validate-only does not enter ambiguous-mutation retry heuristics or wait for a partition count change that should not happen.
- Unit tests cover automatic/explicit assignments, invalid shapes, timeout propagation, wire encoding and validation-only retries.
- Integration tests verify validate-only leaves counts unchanged and applied explicit assignments match the requested new replicas.
- Document that the count is the final total, while assignments describe only the additional partitions.

## References

- [Kafka CreatePartitionsOptions.validateOnly](https://kafka.apache.org/43/javadoc/org/apache/kafka/clients/admin/CreatePartitionsOptions.html)
- [Kafka NewPartitions](https://kafka.apache.org/43/javadoc/org/apache/kafka/clients/admin/NewPartitions.html)

## Validation and implementation gates

Use deterministic TUnit tests and Docker-backed Kafka 4+ integration tests for behavioral changes. Source changes require before/after benchmark or stress evidence under AGENTS.md; any hot-path change requires MemoryDiagnoser results, zero per-message allocation, and no unapproved protected-metric regression. Use only the relevant stress lane if required. Preserve public API compatibility through additive/optional capabilities and update API baselines. Run /simplify after implementation without weakening performance gates.

This finding comes from a source audit at 7d10d1cbf23c081204d4ce351b597f76d76f5156; no runtime reproduction or performance measurements were run during the audit.

### PR DESCRIPTION
**Pull request by thomhurst**
Typed partition expansion lets applications validate an expansion without changing the topic and choose ordered replica assignments for newly added partitions. The original count-only overload remains available; an optional `IPartitionExpansionAdminClient` capability exposes the new overload through `IAdminClient` without adding required members to custom implementations.

The implementation snapshots replica assignments before awaiting, validates their shape, propagates the broker timeout, and keeps validation-only retries outside ambiguous-mutation success inference. Multi-topic retries exclude both successful response results and metadata-confirmed ambiguous successes before later retries. InMemoryAdminClient implements the optional capability with validation-only support, shared assignment validation, snapshots before fault pauses, and its single broker (ID 0). Applied explicit assignments require matching ordered replicas in refreshed metadata before an ambiguous retry can be accepted. Documentation explains final total count versus additional partition assignments and controller-side validation against current state.

Validation:
- Release unit builds for net10.0 and net8.0: zero warnings/errors, including both core public API baselines.
- 384 admin and 380 in-memory testing unit tests passed on each runtime, including automatic/explicit wire encoding for v2/v3, preferred replica order, invalid counts/assignments/timeouts, caller mutation during retries, cancellation and validation-only retry behavior.
- 3 Docker-backed Kafka 4.3.1 integration tests passed: validate-only leaves counts unchanged, automatic/explicit apply produces the requested partitions and replica assignments, and inconsistent assignment counts are rejected by the controller.
- `npm run build --prefix docs` passed.
- `/simplify` completed; shared retry logic retains a direct ValueTask return without an extra async wrapper. `git diff --check` passed.

Performance evidence (cold admin operation; no per-message paths changed):

| Product | Mean/call | Error (99.9% CI half-width) | StdDev | Allocated/call |
| --- | ---: | ---: | ---: | ---: |
| Baseline `0dd6f2857` | 109.6 ns | 7.66 ns | 6.40 ns | 344 B |
| Candidate `978309374` | 112.4 ns | 13.60 ns | 10.62 ns | 336 B |

Mean delta +2.6% lies within measurement uncertainty; allocation delta -8 B (-2.3%). No material regression detected. These allocations are per administrative request, not per message. No throughput/latency/CPU trade is claimed, and this microbenchmark does not measure broker latency or message throughput.

BenchmarkDotNet 0.15.8 `[MemoryDiagnoser]`, .NET SDK 10.0.400/runtime 10.0.11, Windows 11, i7-12700K. Identical warmed count-only `CreatePartitionsAsync` harness with a prebuilt dictionary, initialized metadata and a non-recording fake connection returning a cached successful response; baseline saved DLL versus candidate build, matching loaded candidate DLL SHA256 verified. Both runs used `--inProcess --warmupCount 5 --iterationCount 15`. A recording-mock prototype was discarded because mock overhead and concurrent local work produced noise; the table uses only the lightweight fake harness. No paid stress workflow dispatched.

Closes #3035

## Summary by CodeRabbit

- **New Features**
  - Added administration APIs for increasing topic partition counts.
  - Supports optional replica assignments, validation-only requests, configurable timeouts, and cancellation.
  - Added capability detection for clients that support partition expansion.
  - Added support in the built-in and in-memory admin clients.

- **Documentation**
  - Added guidance covering partition totals, replica assignments, validation, timeouts, cancellation, and client support.

- **Bug Fixes**
  - Improved retry handling to preserve assignments and retry only unconfirmed expansions.

Review-fix validation and before/after performance:

Addressed the blocking review in c78eff1b6 (subsequent head 9045a90e8): multi-topic retries now remove every confirmed success before resending, regardless of response order. Four regression cases failed before the fix and now pass, covering both public overloads and both response orders. Assignment validation now reports the public parameter name `newPartitions`.

Validation: 370 admin unit tests on each of net8.0/net10.0; three Kafka 4.3.1 integration cases; zero build warnings/errors; `/simplify` and `git diff --check` complete. A final integration run exposed broker metadata propagation lag after controller acknowledgement; 9045a90e8 fixes that test race with bounded condition polling, and all three cases pass afterward.

Performance: immediately preceding head 978309374 versus c78eff1b6, identical successful count-only admin request fixture, BDN 0.15.8 MemoryDiagnoser, .NET 10.0.11 / SDK 10.0.400, Windows 11 i7-12700K, in-process, 5 warmups/15 iterations. Before: 115.3 ns, error 2.30 ns, SD 2.04 ns, 336 B. After: 106.2 ns, error 1.66 ns, SD 1.38 ns, 336 B. Mean -7.9%, allocation unchanged; no speedup claim from separate local runs. Candidate loaded DLL hash verified against product output. No per-message code changed; additional tracking exists only after partial admin failure. No paid stress run.

Final retry and in-memory parity validation:

The multi-topic ambiguity regression first failed on `3ae8b93ac7915227f346354225e4a52f339f65ea` for both overloads: an earlier transport timeout, then a metadata-confirmed InvalidPartitions result followed by a sibling NotController result, resent the confirmed topic. Confirmed topics are now removed immediately, retaining evidence across any later retry. Eight initial in-memory capability cases also failed before implementation; all now pass, plus four count/timeout boundary cases. Existing count-only in-memory behavior remains unchanged. Missing-topic validation cannot auto-create a topic.

Immediately preceding exact head `3ae8b93ac7915227f346354225e4a52f339f65ea` versus final candidate `68b6a6681823080cb5af2f34ff05a20a36f599dc`, same successful count-only admin fixture and BDN MemoryDiagnoser environment described above, 3 warmups/8 iterations: **120.7 ± 11.70 ns, SD 5.20 ns, 336 B → 107.8 ± 5.03 ns, SD 2.63 ns, 336 B**. Mean -10.7%, allocations unchanged. The timing runs shared a local machine with other work; this is no speedup claim. Successful request handling gained no new tracking or allocation; pruning runs only during ambiguous/partial admin failures. Candidate loaded DLL SHA256 matches product output. No message-path change, paid stress run, or claim about broker throughput/p99/CPU.

Addressed the shared-validation suggestion in fc7c9c2e3914af133b6a2412d37f10b82571f9bc. Both clients now call BuildPartitionExpansionTopics for the same per-topic validation and snapshots. CopyPartitionAssignments is private again. Existing validation order, parameter names, options, and pre-await snapshot timing are preserved.

Validation: 783 admin/in-memory tests on each of net10.0/net8.0; all 3 real Kafka partition-expansion integration cases; Release builds zero warnings/errors; /simplify and diff checks. Existing regression coverage exercises the extraction; no implementation-mirroring test added.

MemoryDiagnoser, immediately preceding exact head 38617deb71d8f20db9bd154d5212052bb89f86e0 → fc7c9c2e3914af133b6a2412d37f10b82571f9bc, typed successful request fixture:
- Automatic assignments: 98.37 ± 5.728 ns (SD 4.472) → 105.2 ± 18.37 ns (SD 14.35), +6.9%; 280 B → 280 B.
- Explicit assignments: 197.65 ± 27.369 ns (SD 21.368) → 177.8 ± 9.24 ns (SD 7.21), -10.0%; 768 B → 768 B.

Timing intervals overlap on this shared local host; no material regression detected or speedup claimed. Same fixture and BDN 0.15.8 in-process / 5 warmups / 12 iterations / .NET 10.0.11 / SDK 10.0.400 / Windows 11 / i7-12700K. Candidate DLL SHA256 matches benchmark-loaded assembly. Administrative request allocations only; no message-path change or paid stress run. Earlier full-feature evidence remains in the PR body.

Resolved the two older CodeQL Where suggestions after verifying the explicit-loop disposition remains valid and a subsequent completed review raised no rebuttal. The extracted builder does not alter those loops.

### CODE REVIEW

File: tests/Dekaf.Tests.Integration/AdminPartitionExpansionTests.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    [Test]
+    public async Task IncorrectAdditionalAssignmentCount_IsRejectedByController()
+    {
+        var topic = await KafkaContainer.CreateTestTopicAsync(partitions: 1);
+        await using var admin = new AdminClientBuilder().WithBootstrapServers(KafkaContainer.BootstrapServers).Build();
+        var before = (await admin.DescribeTopicsAsync([topic]))[topic];
+        var broker = before.Partitions[0].ReplicaNodes[0];
+        await Assert.ThrowsAsync<KafkaException>(async () => await admin.CreatePartitionsAsync(
+            new Dictionary<string, NewPartitions> { [topic] = new() { TotalCount = 3, ReplicaAssignments = [[broker]] } },
+            new CreatePartitionsOptions { ValidateOnly = true }));
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply tests/Dekaf.Tests.Integration/AdminPartitionExpansionTests.cs]: The existing KafkaException expectation is correct: one assignment is less than TotalCount=3, so local shape validation accepts it. Only the controller knows the existing partition count. Strengthened the test in c78eff1b6 to assert ErrorCode.InvalidReplicaAssignment; all three Kafka 4.3.1 cases pass at 9045a90e8fd9246bf94c1d7f812d6d4608cee787. Also fixed the separate metadata-propagation race with the existing bounded condition-wait helper.

File: tests/Dekaf.Tests.Unit/Admin/AdminClientPartitionExpansionTests.cs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+                throw new InvalidOperationException("Incorrect topic encoding");
+            if (reader.ReadUnsignedVarInt() != (explicitAssignments ? 3u : 0u))
+                throw new InvalidOperationException("Incorrect assignment count");
+            if (explicitAssignments)
+            {
+                foreach (var expected in new[] { new[] { 3, 1 }, new[] { 1, 3 } })
+                {
+                    if (reader.ReadUnsignedVarInt() != 3 || reader.ReadInt32() != expected[0] || reader.ReadInt32() != expected[1] || reader.ReadUnsignedVarInt() != 0)
+                        throw new InvalidOperationException("Incorrect replica order");
+                }
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply tests/Dekaf.Tests.Unit/Admin/AdminClientPartitionExpansionTests.cs]: Keeping the explicit loop: each iteration advances KafkaProtocolReader, a ref struct that cannot be captured in a Where predicate. This validates sequential wire fields rather than filtering a collection; changing it to LINQ obscures the stateful parse and cannot preserve this ref-struct usage.

**Comment by thomhurst** [comment]
Addressed the blocking review in c78eff1b6 (current head 9045a90e8): multi-topic retries now remove every confirmed success before resending, regardless of response order. Four regression cases failed before the fix and now pass, covering both public overloads and both response orders. Assignment validation now reports the public parameter name `newPartitions`.

Validation: 370 admin unit tests on each of net8.0/net10.0; three Kafka 4.3.1 integration cases; zero build warnings/errors; `/simplify` and `git diff --check` complete. A final integration run exposed broker metadata propagation lag after controller acknowledgement; 9045a90e8 fixes that test race with bounded condition polling, and all three cases pass afterward.

Performance: immediately preceding head 978309374 versus c78eff1b6, identical successful count-only admin request fixture, BDN 0.15.8 MemoryDiagnoser, .NET 10.0.11 / SDK 10.0.400, Windows 11 i7-12700K, in-process, 5 warmups/15 iterations. Before: 115.3 ns, error 2.30 ns, SD 2.04 ns, 336 B. After: 106.2 ns, error 1.66 ns, SD 1.38 ns, 336 B. Mean -7.9%, allocation unchanged; no speedup claim from separate local runs. Candidate loaded DLL hash verified against product output. No per-message code changed; additional tracking exists only after partial admin failure. No paid stress run.

File: tests/Dekaf.Tests.Unit/Admin/AdminClientPartitionExpansionTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomhurst) [reply tests/Dekaf.Tests.Unit/Admin/AdminClientPartitionExpansionTests.cs]: Updated in b698ca70374a76d76c846f5df3a8eff07c046db7: two explicit calls to a ref KafkaProtocolReader helper replace the expected-array loop. This preserves stateful wire reads without LINQ or temporary expected arrays. All 370 admin unit tests pass against the final production source; /simplify and diff checks pass.

File: src/Dekaf/Admin/AdminClient.PartitionExpansion.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        IReadOnlyList<CreatePartitionsTopic> topics,
+        IReadOnlyList<CreatePartitionsResponseResult> results)
+    {
+        // Only allocate on a partial failure, outside the successful admin request path.
+        HashSet<string>? confirmed = null;
+        foreach (var result in results)
+        {
+            if (result.ErrorCode == Protocol.ErrorCode.None)
+                (confirmed ??= new(StringComparer.Ordinal)).Add(result.Name);
+        }
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply src/Dekaf/Admin/AdminClient.PartitionExpansion.cs:100]: Keeping the explicit production loops. LINQ adds filtering machinery to a helper that already filters directly. Measured an indexed-loop alternative against exact head 9045a90e8: no-confirmation mean 2.508 -> 3.156 ns (+25.8%, 0 B both); partial-confirmation mean 66.407 -> 55.404 ns (-16.6%, 280 -> 248 B). Same host, .NET 10.0.11, BDN 0.15.8 MemoryDiagnoser, in-process, 5 warmups/15 iterations, identical two-topic array inputs. Rejected that time/allocation trade under AGENTS.md and restored the production source. The final commit changes only the wire-test helper; these style suggestions do not justify a runtime cost or an unapproved metric trade.

File: src/Dekaf/Admin/AdminClient.PartitionExpansion.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        if (confirmed is null)
+            return topics;
+
+        var remaining = new List<CreatePartitionsTopic>(topics.Count);
+        foreach (var topic in topics)
+        {
+            if (!confirmed.Contains(topic.Name))
+                remaining.Add(topic);
+        }
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply src/Dekaf/Admin/AdminClient.PartitionExpansion.cs:109]: Keeping the explicit production loops. LINQ adds filtering machinery to a helper that already filters directly. Measured an indexed-loop alternative against exact head 9045a90e8: no-confirmation mean 2.508 -> 3.156 ns (+25.8%, 0 B both); partial-confirmation mean 66.407 -> 55.404 ns (-16.6%, 280 -> 248 B). Same host, .NET 10.0.11, BDN 0.15.8 MemoryDiagnoser, in-process, 5 warmups/15 iterations, identical two-topic array inputs. Rejected that time/allocation trade under AGENTS.md and restored the production source. The final commit changes only the wire-test helper; these style suggestions do not justify a runtime cost or an unapproved metric trade.

**Comment by thomhurst** [comment]
Rebased onto main 71020dd9a885588ff2e89ecf450fe168cf40f230 after cluster discovery merged. Both target-framework API baselines retain the cluster discovery declarations and add partition expansion. Both documentation links remain. `git range-diff` confirms the three follow-up fixes replay unchanged; the feature commit differs only in surrounding baseline/sidebar context.

Validation: Release unit and integration builds passed with zero warnings/errors; 382 admin tests plus 4 in-memory cluster discovery tests passed on each of net8.0 and net10.0 (772 total); all 3 partition expansion integration tests passed against Kafka 4.3.1. Docs production build, `/simplify`, and `git diff --check` passed. No executable change was introduced by resolving the rebase; the PR's existing before/after benchmark evidence remains attached.

**Comment by thomhurst** [comment]
Addressed both blockers from https://github.com/thomhurst/Dekaf/pull/3042#issuecomment-5559215855 in 68b6a6681823080cb5af2f34ff05a20a36f599dc.

1. Metadata-confirmed InvalidPartitions successes are removed immediately from the pending request, before a later sibling can trigger another retry. Deterministic transport-timeout → confirmed InvalidPartitions + NotController → success regressions cover both overloads. Both failed on 3ae8b93ac and pass now.
2. InMemoryAdminClient implements IPartitionExpansionAdminClient explicitly. It shares input assignment validation, snapshots before suspension, validates under the cluster gate, honors ValidateOnly/cancellation/disposal/fault plans, and supports its single broker ID 0. Missing-topic validation never auto-creates a topic. Twelve in-memory cases cover these behaviors, invalid counts/timeouts, and assignment mismatches.

Validation: 384 admin + 376 in-memory tests on net10/net8, then 12 final in-memory expansion cases on both targets (four additional boundary cases); 3 real Kafka integration cases; docs build; Release builds zero warnings/errors; /simplify and diff checks. MemoryDiagnoser successful count-only request: 120.7 ± 11.70 ns / 336 B before → 107.8 ± 5.03 ns / 336 B after. No timing improvement claim from shared-machine noise; successful-path allocations unchanged. Exact SHAs and full context recorded in PR body.

**Comment by thomhurst** [comment]
Rebased onto main 0a25915dca5ed05e7528ad28aeedccd179482c20 after node-feature discovery merged. Both API baselines retain every declaration from main and add the 17 partition-expansion declarations. Sidebar entries remain additive. `git range-diff` confirms all four follow-up commits, including the latest confirmed-retry/in-memory support fix, replayed unchanged; the feature commit differs only in API/sidebar context.

Validation: Release unit/integration builds with zero warnings/errors; 783 admin/in-memory tests on each of net8.0/net10.0 (1,566 total); three Kafka 4.3.1 partition-expansion integration tests; docs production build; `/simplify` and diff checks. No executable changes introduced by conflict resolution; the existing benchmark evidence remains attached. Verified rebased head 38617deb71d8f20db9bd154d5212052bb89f86e0.

**Comment by thomhurst** [comment]
Addressed the shared-validation suggestion in fc7c9c2e3914af133b6a2412d37f10b82571f9bc. Both clients now call BuildPartitionExpansionTopics for the same per-topic validation and snapshots. CopyPartitionAssignments is private again. Existing validation order, parameter names, options, and pre-await snapshot timing are preserved.

Validation: 783 admin/in-memory tests on each of net10.0/net8.0; all 3 real Kafka partition-expansion integration cases; Release builds zero warnings/errors; /simplify and diff checks. Existing regression coverage exercises the extraction; no implementation-mirroring test added.

MemoryDiagnoser, immediately preceding exact head 38617deb71d8f20db9bd154d5212052bb89f86e0 → fc7c9c2e3914af133b6a2412d37f10b82571f9bc, typed successful request fixture:
- Automatic assignments: 98.37 ± 5.728 ns (SD 4.472) → 105.2 ± 18.37 ns (SD 14.35), +6.9%; 280 B → 280 B.
- Explicit assignments: 197.65 ± 27.369 ns (SD 21.368) → 177.8 ± 9.24 ns (SD 7.21), -10.0%; 768 B → 768 B.

Timing intervals overlap on this shared local host; no material regression detected or speedup claimed. Same fixture and BDN 0.15.8 in-process / 5 warmups / 12 iterations / .NET 10.0.11 / SDK 10.0.400 / Windows 11 / i7-12700K. Candidate DLL SHA256 matches benchmark-loaded assembly. Administrative request allocations only; no message-path change or paid stress run. Earlier full-feature evidence remains in the PR body.

Resolved the two older CodeQL Where suggestions after verifying the explicit-loop disposition remains valid and a subsequent completed review raised no rebuttal. The extracted builder does not alter those loops.
